### PR TITLE
Support instance-as-result children

### LINKED ISSUE #19: Could not handle instance-as-result from Stateless component
Original issue - https://github.com/gaearon/react-hot-loader/issues/832
React-Hot-Loader is a bit hacky thing, and it's functional component wrapper will return an `instance` from render.
This is a not common, but absolutely legit behavior, called `IndeterminateComponent`. Used by Relay as I know (https://github.com/facebook/relay/blob/d77a34c6b5abddfe078b0d50c2d43181fe35d574/packages/react-relay/classic/container/RelayContainer.js#L1050-L1055, dont know anyone else).
Unfortunately react-tree-walker(React.Children.count, to be more correct) cannot stand this, throwing an error.

Example - https://codesandbox.io/s/6wz0or0rqk, or in original issue.

### PR DESCRIPTION
**Pull request by dvkndn**
Fixes #19

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -69,24 +69,34 @@ export default function reactTreeWalker(
...
 
         const child = getChildren()
+        // This is to support instance-as-result case. In other words: sometimes
+        // the returned value is not the child itself but a function to render
+        // that child.
+        // See:
+        // - https://github.com/ctrlplusb/react-tree-walker/issues/19
+        // - https://github.com/gaearon/react-hot-loader/issues/832
+        // - https://github.com/gaearon/react-hot-loader/blob/master/src/proxy/utils.js#L84
+        const theChild = child && typeof child.render === 'function'
```
Comment (theKashey) [inline src/index.js]: Could one intermediate component return another? 
Maybe this should be recursive.

**Comment by theKashey** [comment]
So - in case of "instance" in tree - just "skip" it by rendering, and dive deeper.
Should work, but I'll and a test for this case. Just to be sure.

File: src/index.js (hunk 1 above)
Comment (dvkndn) [reply src/index.js]: Do you mean something like this:

```js
const getFinalChild = component => (
  component && typeof component.render === 'function'
    ? getFinalChild(component.render())
    : component
);

const theChild = getFinalChild(child);
```

or

```js
let child = getChildren();

while (child && typeof child.render === 'function') {
  child = child.render();
}
```

I don't know what to name the function yet btw...

Comment (theKashey) [reply src/index.js]: I'll prefer the first one. It is just testable, and look like code coverage is a thing here.

Comment (dvkndn) [reply src/index.js]: I'll update shortly. Meanwhile, if you have time, can you help me on having some tests for this? Like how to have a sample child with `render`... Maybe  something like this?

```jsx
const Foo = () => (
  () => <span>abc</span>
);
```

Comment (theKashey) [reply src/index.js]:
```js
class MyComponent extends Component { render() { return <div /> }}

const Foo = (props) => new MyComponent(props);
```

**Comment by dvkndn** [comment]
@theKashey I added a test and add support for recursive calling render until we have a valid React child.

just noticed that I mistakenly commited built files.. I should exclude these from the PR, right @ctrlplusb ?

**Comment by ctrlplusb** [comment]
Thanks so much for this @dvkndn! And for @theKashey and @birkir for the review 👍

**Pull request merged**